src/Scrutor/ServiceCollectionExtensions.Decoration.cs

```diff
@@ -287,8 +287,11 @@ private static bool TryDecorateDescriptors(this IServiceCollection services, Typ
             {
                 var index = services.IndexOf(descriptor);
 
+                // replace original descriptor with implementation type
+                services[index] = ReplaceWithImplementation(descriptor);
```

Comment:
Replacing a descriptor with the implementation type may break existing registrations

```
// Before registration of the decorator
ServiceCollection services = ...;

services.AddTransient<Foo, Foo2>();
services.AddTransient<IFoo, Foo>();
	
Resolve<IFoo>  -> Foo
Resolve<Foo>  -> Foo2
```

Now we will decorate `IFoo` with `FooDecorator`

```
// the decoration will replace
services.AddTransient<IFoo, Foo>();
// with
services.AddTransient<Foo, Foo>();
// this breaks the following registration
services.AddTransient<Foo, Foo2>();

Resolve<Foo>  => Foo instead of Foo2
```

Reply:
1. DI lib must be reliable. There will be different kinds of registrations in the wild, especially if we trying to decorate a complex 3rd party lib (e.g. EF Core / ASP.NET Core). 
2. It does :P 
   try the following

```
var services = new ServiceCollection();
	
services.AddTransient<Foo, Foo2>();
services.AddTransient<IFoo, Foo>();
	
//services.Decorate<IFoo,FooDecorator>();
	
var provider = services.BuildServiceProvider();
```

without `Decorate`
```
provider.GetRequiredService<IFoo>().Dump();  // Foo
provider.GetRequiredService<Foo>().Dump();   // Foo2
```

with `Decorate`

```
provider.GetRequiredService<IFoo>().Dump();  // FooDecorator+Foo
provider.GetRequiredService<Foo>().Dump();   // Foo instead of Foo2
```

Reply:
1. Absolutely agreed, my argument is that the current behaviour _isn't_ "reliable" - disposable services aren't disposed by the container when you use the current implementation. 
2. True, but you cheated in that example and moved the `Foo2` registration up before the decorate call, it was after it previously 😛

I still agree that there's too many sharp edges for this to be acceptable. I think it's worth pursuing though. 🙂